model/project.go

```diff
@@ -92,13 +93,19 @@ type BuildVariantTaskUnit struct {
 	// Name has to match the name field of one of the tasks or groups specified at
 	// the project level, or an error will be thrown
 	Name string `yaml:"name,omitempty" bson:"name"`
-	// IsGroup indicates that it is a task group or a task within a task group.
-	// This is always populated after translating the parser project to the
-	// project.
+	// IsGroup indicates that it is a task group. This is always populated for
+	// task groups after translating the parser project to the project.
 	IsGroup bool `yaml:"-" bson:"-"`
-	// GroupName is the task group name if this is a task in a task group. If
-	// it is the task group itself, it is not populated (Name is the task group
-	// name).
+	// IsPartOfGroup indicates that this unit is a task within a task group. If
+	// this is set, then GroupName is also set. Note that project translation
+	// does not expand task groups into their individual tasks, so this is only
+	// set for special functions that explicitly expand task groups into
+	// individual task units (such as FindAllBuildVariantTasks).
+	IsPartOfGroup bool `yaml:"-" bson:"-"`
+	// GroupName is the task group name if this is a task in a task group. This
```

Comment:
If this field is only ever set if `IsPartOfGroup` is true, would it be possible to instead  remove the `IsPartOfGroup` field entirely and instead implement a function like:
```
func (bvt *BuildVariantTaskUnit) IsPartOfGroup() bool {
	return bvt.GroupName != ""
}
``` 
instead?

Reply:
This would also work, but I considered that we could (legitimately) change the behavior of `GroupName` one day so that for convenience, it's set for both when it's part of a task group or when it's the task group itself, which would break the distinction of task group vs. part of a task group.

I did also consider just not setting `IsGroup` + `GroupName` at all if it's part of a group (a task unit being part of a group is only relevant in [this one spot](https://github.com/evergreen-ci/evergreen/blob/main/model/lifecycle.go#L1242-L1246)), but it could make it difficult to use `CreateTasksFromGroup` and `FindAllBuildVariantTasks` in the future if we can't distinguish a regular task from one in a task group.

Reply:
Sounds good, makes sense if you want the `GroupName` to potentially take on a new meaning in the future. As for removing the `IsGroup` + `GroupName` entirely, the last solution you brought up sounds the most elegant, but I can see how it would be a bigger refactor as you said. I'm fine with how this works in this PR, and I'll leave it to you if you want to make a followup ticket to remove the fields.